[LRN] Add InertMath: non-editable with no mouse events or cursor

karoltarasiuk opened the pull request:
Used to render keyboard buttons so that they don't hijack
focus when they're mousdown'ed.

Conflicts:
    src/commands/math.js

laughinghan commented on src/commands/math.js:
```diff
@@ -466,3 +466,23 @@ API.MathField = function(APIClasses) {
+/**
+ * This is almost the same as StaticMath, except we don't
+ * bind any mouse events (which give us selectability and move
+ * around an invisible cursor).
```
ah, this makes sense

laughinghan commented on src/commands/math.js:
```diff
@@ -466,3 +466,23 @@ API.MathField = function(APIClasses) {
+    this.RootBlock = MathBlock;
+    _.__mathquillify = function() {
+      super_.__mathquillify.call(this, 'mq-math-mode');
+      this.__controller.staticMathTextareaEvents();
```
as far as I know, `staticMathTextareaEvents()` is part of making mouse selection of StaticMath's work (that is, after mouse selecting, it lets you Cmd-C to copy the selection), and doesn't do anything else, so I don't think you even need this

karoltarasiuk replied:
makes sense updating

karoltarasiuk replied:
Actually I could remove this, but it wasn't fixing selectability issues. And I noticed that I shouldn't inferit from StaticMath but from AbstractMathQuill, and then it works.

laughinghan commented on test/demo.html:
```diff
@@ -59,6 +59,18 @@
+
+<p>
+  But if you won't use InertMath, the result will be far from perfect:<br>
+  Insert <button class="insert-trigger" data-latex="\sqrt{4}"><span class="mathquill-math-field">\sqrt{4}</span></button> inside formulas below:<br>
```
Why do you have an editable math field inside a button?

karoltarasiuk replied:
Removed this - was rubbish indeed. I edited previous example to make it more meaningful.

laughinghan commented on test/demo.html:
```diff
@@ -116,6 +129,14 @@
+  $(this).parent().find('.mathquill-math-field').each(function () {
+    var mathquill = MQ(this);
+    mathquill.write(latex);
+  });
```
this appears to seek out every editable math field in the paragraph and write `\sqrt{4}` into it, including the editable math field in the button, is this not expected behavior?

karoltarasiuk replied:
^^^

laughinghan commented on test/demo.html:
```diff
@@ -59,6 +59,18 @@
 <p>This button runs the JavaScript code written on it to MathQuill-ify the following <code>&lt;span&gt;</code> element into an editable math textbox: <button onclick="$(this).text('MQ(this.nextSibling).revert()'); MQ.MathField(this.nextSibling); var orig = arguments.callee; this.onclick = function(){ $(this).text('MQ.MathField(this.nextSibling)'); MQ(this.nextSibling).revert(); this.onclick = ori ...
 
+<p>
+  Insert <button class="insert-trigger" data-latex="\sqrt{4}"><span class="mathquill-inert-math">\sqrt{4}</span></button> inside formulas below:<br>
```
Using a StaticMath (`mathquill-static-math`) seems to work equally as well as an InertMath in this case?

karoltarasiuk replied:
Yep, see above - I made a mistake and correct it now.

laughinghan commented:
I think this could be a useful config option on static math instances, but I don't think this ought be its own top-level mathquill API class, and static math also doesn't seem to have the problem you're fixing (perhaps anymore)?

karoltarasiuk commented:
@laughinghan added changes, as really it wasn't working as expected because of a bug in the code. Now it works as expected and you can really see it in the demo page.

Static math is selectable, inert is not - I'd still argue that both are useful. Could introduce a flag like `ignoreInteractions` and then load StaticMath with a lot of `if`s in there. Don't think it would be ideal as now code looks very neat and tidy. But your call after all.

laughinghan commented:
Actually I like this so much it should be the default behavior for StaticMath in the first place! :) It's basically what KaTeX does.

I think it's actually just one `if` statement, the other thing that StaticMath does that InertMath doesn't is `.innerFields`, which I don't think you're deliberately disabling, right?

Since it's a breaking change I'll submit a little PR to add it using the API versioning scheme

kschaefe commented:
We use StaticMath with `.innerFields`.  Just confirming that you're not removing that functionality.

laughinghan commented:
I'm not! I'm not removing `.innerFields`, I think that's an important feature that InertMath needs too

kschaefe commented:
Great! It was late last night and I think my paranoia must have been getting the best of me.

karoltarasiuk commented:
@laughinghan I've added `.innerFields` on a separate commit, but I don't think they're valid use case here... I've added changes to the demo example as well, and you can see, that having `.innerFields` in InertMath can be rather pointless. Unless you don't want them to be editable, but only renderable.

kschaefe commented:
We just came across a bug in our software that would be solved by this.  What's the status on getting this merged?

laughinghan commented:
Well I don't want a separate `InertMath`, I want this to be a config option on `StaticMath`.

Also,

> having `.innerFields` in InertMath can be rather pointless. Unless you don't want them to be editable, but only renderable.

this seems surprising to me, this is def not what I'd want, what I'd expect is for the inner fields to be editable, but selecting the inert math does normal browser selection rather than MathQuill selection.

Maybe they appear inert because they don't respond to clicks, because inner editable fields in an inert static math need to notice it's inert and call `.delegateMathFields()` on themselves (the inner editable fields)?
